[IMP] mrp_bom_version: Show in tree buttons draft, activate, new vers…

alfredoavanzosc opened the pull request:
…ion, and historical. Create _copy_bom, _update_bom_state_after_copy functions.

[ADD] mrp_bom_new_version_when_component_change: Create a new version of BoM when are you going to change a component.

mrp_bom_version: Módulo modificado para poder mostrar los botones del formulario en el tree. Además se han creado 2 funciones nuevas, para luego poder heredarlas en el siguiente módulo nuevo.
mrp_bom_new_version_whn_componen_change: Nuevo módulo para crear una nueva versión de la LdM a la cuál se le va a cambiar el componente de una forma masiva.

alfredoavanzosc committed:
[IMP] mrp_bom_version: Show in tree buttons draft, activate, new version, and historical. Create _copy_bom, _update_bom_state_after_copy functions.
[ADD] mrp_bom_new_version_when_component_change: Create a new version of BoM when are you going to change a component.

alfredoavanzosc commented:
Se me había olvidado comentar que este nuevo requerimiento esta solicitado en la tarea Tarea CLM0072 - Versionado LMs en cambios masivos de componente

pedrobaeza commented:
No entiendo por qué esto se ha hecho como nuevo módulo, y no como parte del módulo ya existente de cambio masivo de componentes. Debería formar parte de él, y si acaso, desde el propio asistente, incluir una marca que indique si se quiere hacer versión o no, e incluso si al hacer la versión, se desea que la anterior se historifique o no, que fue otra de las reclamaciones según recuerdo.

anajuaristi commented:
@pedrobaeza no se ha hecho nuevo módulo bom_version es anterior al de cambio masivo de componentes y es un módulo ya estable usado por varios clientes.
Lo de historificar, se puso obligatorio que la anterior se historificase y está perfecto así. Si no quieren historificar, pueden sacar primero una copia y luego historificar. Por mi parte está perfecto así y no lo tocaría
Lo único que le comenté a Alfredo de sacar los botones y estado del bom al tree para que dé más visibilidad y se pueda ejecutar desde ahí, pero nada más.

anajuaristi commented:
@pedrobaeza entiendo que te refieresa al mrp_bom_new_version_when_component_change. Estoy de acuerdo en que no se meta nuevo módulo. 
Simplemente bastaría con el anterior, y meterle dependencia con el bom_version. Sin más. 
@alfredoavanzosc lo hacemos así, porfa.

alfredoavanzosc committed:
[IMP] mrp_bom_component_change: Introduce programming module mrp_bom_new_version_when_component_change.

alfredoavanzosc commented:
Realizado lo solicitado por Ana, y por Pedro, y subido el módulo "mrp_bom_component_change" modificado.

pedrobaeza commented on mrp_bom_component_change/README.rst:
```diff
@@ -1,4 +1,7 @@
-This module allows to change massively one component by another on a list of BoMs.
+This module allows to change massively one component by another on a list of
+BoMs. Before you compose a massive change in BoMs, the BoM is to be changed It
+will state "Historical", and changing component is made in a new version of the
```
Pero no hemos dicho de incluir casillas para indicar si se hace o no ambas cosas?

anajuaristi replied:
No... para eso hay que cambiar bom_version y no tiene sentido. Si no quieren historificar que saquen una copia y listo. 
Bom_version ya está estable y nadie ha pedido eso salvo una única persona. Además se puede cubrir facilmente copiando antes de ejecutar el cambio, si lo necesitan. Yo no metería más código aquí.

pedrobaeza replied:
Que no hay que cambiar nada. Es poner un par de ifs en el código.

pedrobaeza commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -65,6 +65,21 @@ def do_component_change(self):
+        bom_ids = []
+        for bom in self.boms:
+            change_bom = False
+            for bom_line in bom.bom_line_ids:
```
Hazlo con filtered: `bom.bom_line_ids.filtered(lambda x: x.product_id == self.old_component.id)`

pedrobaeza commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -65,6 +65,21 @@ def do_component_change(self):
+                    change_bom = True
+                    break
+            if not change_bom:
+                bom_ids.append(bom.id)
```
Para qué lo añades a bom_ids?

anajuaristi commented:
@alfredoavanzosc 
Meteremos un nuevo check en la pantalla principal mrp_bom_component_change
"create new version"

Si este check se marca, al realizar el cambio se ejecutará la función new versión pero si no está marcado se realizará únicamente el cambio de componente, tal cual se hacía antes y no se creará nueva versión.

alfredoavanzosc committed:
[IMP] mrp_bom_component_change: If you checked the new fied "Create new BoM version", before you compose a massive change in BoMs, the BoM is to be changed It will state "Historical", and changing component is made in a new version of the BoM.

alfredoavanzosc commented:
Realizadas las mejoras solicitadas por Pedro, y subido el módulo.

pedrobaeza commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -48,10 +51,11 @@ def onchange_operation(self):
+                    lambda x: x.product_id.id == self.old_component.id)
+                if bom_lines:
+                    bom_lst.append(bom.id)
+                    break
```
Este break aquí está mal. Corta el bucle.

pedrobaeza commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -48,10 +51,11 @@ def onchange_operation(self):
+                bom_lines = bom.bom_line_ids.filtered(
+                    lambda x: x.product_id.id == self.old_component.id)
+                if bom_lines:
+                    bom_lst.append(bom.id)
```
Haz un recordset en lugar de una lista de ids. Esto ya lo hemos hablado muchas veces.

pedrobaeza commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -48,10 +51,11 @@ def onchange_operation(self):
+                if bom_lines:
+                    bom_lst.append(bom.id)
+                    break
             self.boms = bom_lst
```
Por qué almacenas en self.boms esta lista/recordset?

pedrobaeza replied:
Ah, OK, ya he visto que lo haces en el onchange.

pedrobaeza replied:
Deberías utilizar un campo calculado, no un onchange.

alfredoavanzosc replied:
A ver Pedro... todo lo que me estas poniendo que modifique no lo he hecho yo, viene de antes. Yo lo cambio, no hay ningún problema. ¿Pero como pongo esto como campo calculado?, al principio si que valdría, teniendo dependencia con el componente viejo, ¿pero luego como lo hago, si tengo que sustituir porque se requiere historificar?

pedrobaeza replied:
Yo no sé de quién viene, Alfredo. Lo estoy revisando ahora y lo veo mejor así (aunque en el pasado se me pasara revisarlo). No entiendo tu pregunta, simplemente pones el `@api.depends('old_component')` y no hay nada más que hacer.

alfredoavanzosc replied:
Eso esta claro, y hay tendré las LdM que tengo que cambiar, ¿Pero como lo hago si tienen marcado el nuevo campo de crear nueva version en las LdM a tratar?, porque en este campo que quieres que sea ahora calculado en un principio se mostraran las LdM con el componente viejo a cambiar, pero si marco el nuevo campo de crear nueva versión de la LdM, dejaré de mostrar estas, y tendré que mostrar las nuevas. ¿Como hago esto si es un campo calculado?

pedrobaeza replied:
Alfredo, no te líes, y céntrate en lo que se quiere: calcular las LdMs afectadas por el cambio de componente. Si en boms lo que estás haciendo en visualizar el resultado, cosa que tampoco veo correcto, ya que normalmente lo que se hace es que cuando se termina el asistente, devuelves un diccionario con los ids de los elementos a mostrar y te lleva a la vista lista de esos elementos (como al generar facturas), pero si ése fuera el caso, créate otro campo calculado, pero ya te digo que lo que se pretende es eso, calcularlo una vez.

Según yo entendía el desarrollo (y sin haberlo probado), el campo boms era una previsualización de las LdMs a modificar, y por eso te decía de hacerlo calculado.

Con todo esto que te he dicho, rehaz lo que haya que rehacer para que funcione así.

pedrobaeza commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -65,10 +69,21 @@ def do_component_change(self):
+        if self.create_new_version:
+            bom_ids = self.boms
+            for bom in bom_ids:
+                bom_lines = bom.bom_line_ids.filtered(
```
Utiliza la lista ya calculado del campo que te he dicho.

pedrobaeza replied:
No hace falta volver a comprobar si existe el componente

alfredoavanzosc committed:
[IMP] mrp_bom_component_change: Improve programming.

alfredoavanzosc commented:
Realizadas las mejoras solicitadas por Pedro, y subido el módulo modificado.

oihane commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -66,9 +68,18 @@ def do_component_change(self):
-                    bom_line.product_id = self.new_component.id
+            bom_lines = bom.bom_line_ids.filtered(
+                lambda x: x.product_id.id == self.old_component.id)
+            if bom_lines:
```
Porque este `if`, se supone que solo vas a procesar aquellas listas de materiales que tengan el componente que quieres cambiar.

alfredoavanzosc replied:
Porque no solo crea una nueva LdM a la cuál realizará los cambios, sino que también modifica el nuevo componente a las líneas seleccionadas.

pedrobaeza replied:
Lo que Oihane dice es que no necesitas poner el if, ya que es seguro que se van a encontrar líneas. Es más, no deberías hacer la búsqueda si no es el `else` del `if self.create_new_version:`

oihane commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -83,5 +94,4 @@ def do_revert(self):
 
     @api.one
     def clear_list(self):
```
@pedrobaeza es un campo calculado tiene sentido que se siga manteniendo?

alfredoavanzosc replied:
Tienes toda la razón, es un campo calculado que depende de la introducción del componente viejo, pero este campo calculado es editable, se pueden quitar o añadir líneas, y he dejado el botón de limpiar la lista de LdMs a tratar, que ya existía antes, si hay que quitarlo..... lo quito.

pedrobaeza replied:
Un campo calculado no es escribible a no ser que tenga una función inversa, así que habría que quitarlo, sí

pedrobaeza commented on mrp_bom_component_change/models/mrp_bom_change.py:
```diff
@@ -27,37 +27,39 @@ class MrpBomChange(models.Model):
+                    lambda x: x.product_id.id == self.old_component.id)
+                if bom_lines:
+                    self.boms = [(4, bom.id)]
+            if self.state != 'process':
```
Esto no debes cambiarlo en un campo calculado.

alfredoavanzosc replied:
Buenos días Pedro. El campo "state" no es un campo calculado, antes se ponía en estado "process" en el onchange de componente viejo a cambiar. ¿Que hago volver a poner el onchange de componente viejo para modificar el estado?

pedrobaeza replied:
Ese campo para qué sirve? Yo diría que no aporta nada en el estado actual, no?

alfredoavanzosc replied:
Bueno... empieza con draft, cuando se introduce el componente viejo a cambiar se pone en "process" y te muestra el botón para realizar el cambio. Cuando realiza el cambio pone el estado en "done" y te muestra el botón para deshacer el cambio de componente viejo por nuevo.

pedrobaeza replied:
El primer cambio se puede evitar simplemente poniendo como required el componente viejo y el nuevo (que es lo que debería ser por otro lado y eliminar esa comprobación que hay por ahí). El siguiente estado, como te decía en un comentario anterior, cuando se pulse el botón "Cambiar" o "Procesar" o como se llame, hay que hacer el proceso, y devolver un diccionario para que te cargue la vista con las BoMs modificadas/creadas. Lo de deshacer no tiene sentido ahora que se crea nueva versión, no crees?

anajuaristi commented:
Lo de deshacer no tiene sentido si se ha cogido la opción de versionar/historificar las listas antiguas pero si no se coge esa opción tiene el mismo sentido que tenía antes.
En cualquier caso, no veo mal quitarlo porque siempre pueden volver a seleccionar el componente nuevo y ejecutar de nuevo el cambio poniendo el anterior. Quitamos código y no se pierde mucha funcionalidad. Queda más sencillo.
Por mi parte OK. :+1:

alfredoavanzosc committed:
[IMP] mrp_bom_component_change: Remove the field "state" of the model "Mrp BoM Component Change".

alfredoavanzosc commented:
He quitado el campo estado del modelo "Mrp BoM Component Change". Si Pedro me da ya el OK, me pongo a realizar la carpeta "tests" para los módulos "mrp_bom_version", y "mrp_bom_component_change".

pedrobaeza commented:
Puedes empezar a hacer los tests ya si quieres. Luego reviso todo junto.

alfredoavanzosc committed:
[IMP] mrp_bom_version: Include folder "tests".
[IMP] mrp_bom_component_change: Include folder "tests", and programming improvements.

alfredoavanzosc commented:
Realizadas las mejoras de programación solicitadas por Pedro, y creada la carpeta "tests" para los módulos "mrp_bom_version", y  "mrp_bom_component_change".

pedrobaeza commented:
Muchas gracias por los tests, Alfredo! Ya hemos conseguido un 1% más de cobertura. Vamos a por el 100%, jeje.

:+1:
